### PR TITLE
- Adjust ticker times to show correct time on config changes for custom-ticker-delay

### PR DESCRIPTION
**Pull request by simon44556**
- Adjust ticker execution steps based on custom-ticker-delay changes

## Description

This PR adjust the tick rate based on the config value for custom-ticker-delay.
It will try to keep the tickers execution at a similar speed as to when the default value is used.

I also adjust the displayed time based on the difference of execution time.

## Proposed changes

Changed CraftingOperation to take into account the new ticks config. Normalize the value to the default one of one operation per 10 ticks. 

Changed NumberUtils. Added logic that calculates the time new changes will take to finish the operation.

## Related Issues (if applicable)

Not sure if there is na issue open.

Here is how to reproduce the issue:
1. Adjust custom-ticker-delay to 20.
2. Run a ticker ingame.
3. The display timer will be wrong the execution slows down by 1/2 so the secounds and minutes are not synced to the actual values.

## Checklist

- [x] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [ ] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [ ] I have made sure that the proposed changes do not break compatibility across the supported Minecraft versions (1.14.* - 1.19.*).
- [x] I followed the existing code standards and didn't mess up the formatting.
- [x] I did my best to add documentation to any public classes or methods I added.
- [x] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/NumberUtils.java [hunk 1]
```diff
@@ -171,9 +176,31 @@ private NumberUtils() {}
...
 
-    public static @Nonnull String getTimeLeft(int seconds) {
+    public static @Nonnull String getTimeLeft(int ticksLeft) {
         String timeleft = "";
 
+        /**
+         * This is the amount calculated in addProgress(). 
+         * This will be the scale used for timer.
+         */
+        double timeScale = (TICK_RATE / 10.0d) * 1;
```
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/NumberUtils.java:186]: Is the * 1 necessary here. it will always stay the same.

Comment (simon44556) [reply src/main/java/io/github/thebusybiscuit/slimefun4/utils/NumberUtils.java]: You are correct this is not required. I removed it now.

Comment (J3fftw1) [reply src/main/java/io/github/thebusybiscuit/slimefun4/utils/NumberUtils.java]: i was doubting my self for a tad. opened my calc and did 0.1*1

Comment (svr333) [reply src/main/java/io/github/thebusybiscuit/slimefun4/utils/NumberUtils.java]: lol

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/NumberUtils.java [hunk 2]
```diff
@@ -25,6 +25,10 @@
  *
  */
 public final class NumberUtils {
+    /**
+     * Save config value for tick rate
+     */
+    private static final int TICK_RATE = Slimefun.getCfg().getInt("URID.custom-ticker-delay");
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/NumberUtils.java:31]: This is not safe to call in a static context.

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/NumberUtils.java [hunk 3]
```diff
@@ -35,7 +39,8 @@ public final class NumberUtils {
     /**
      * We do not want any instance of this to be created.
      */
-    private NumberUtils() {}
+    private NumberUtils() {
+    }
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/NumberUtils.java:43]: Please restore the original formatting here

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/NumberUtils.java [hunk 4]
```diff
@@ -171,9 +176,31 @@ private NumberUtils() {}
...
+            seconds = (int) (ticksLeft * (normalScale / 2 + 1.0d));
+        }
+        // Adjust for negative change 
+        else if (normalScale > 0) {
+            seconds = (int) (ticksLeft * (normalScale + 1.0d));
+        }
+        // No change 
+        else {
+            seconds = ticksLeft;
+        }
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/NumberUtils.java:202]: Please adhere to our code style.

```suggestion
        if (normalScale < 0) {
            //Adjust for positive change
            seconds = (int) (ticksLeft * (normalScale / 2 + 1.0D));
        } else if (normalScale > 0) {
            // Adjust for negative change 
            seconds = (int) (ticksLeft * (normalScale + 1.0D));
        } else {
            // No change 
            seconds = ticksLeft;
        }
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/NumberUtils.java [hunk 5]
```diff
@@ -171,9 +176,31 @@ private NumberUtils() {}
...
 
-    public static @Nonnull String getTimeLeft(int seconds) {
+    public static @Nonnull String getTimeLeft(int ticksLeft) {
         String timeleft = "";
 
+        /**
+         * This is the amount calculated in addProgress(). 
+         * This will be the scale used for timer.
+         */
+        double timeScale = (TICK_RATE / 10.0d);
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/NumberUtils.java:186]:
```suggestion
        double timeScale = TICK_RATE / 10.0D;
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/NumberUtils.java [hunk 6]
```diff
@@ -171,9 +176,31 @@ private NumberUtils() {}
...
+        double normalScale = ((timeScale) - Math.round(timeScale));
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/NumberUtils.java:187]: Redundant braces.
```suggestion
        double normalScale = timeScale - Math.round(timeScale);
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/operations/CraftingOperation.java [hunk 7]
```diff
@@ -35,12 +37,21 @@ public CraftingOperation(@Nonnull ItemStack[] ingredients, @Nonnull ItemStack[]
         this.ingredients = ingredients;
         this.results = results;
         this.totalTicks = totalTicks;
+
+        this.tickRate = Slimefun.getCfg().getInt("URID.custom-ticker-delay");
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/operations/CraftingOperation.java:41]: Reading a config setting in a constructor here is not efficient...

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/operations/CraftingOperation.java [hunk 8]
```diff
@@ -21,6 +22,7 @@ public class CraftingOperation implements MachineOperation {
     private final ItemStack[] results;
 
     private final int totalTicks;
+    private int tickRate;
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/operations/CraftingOperation.java:25]: Why is this not final?

Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/operations/CraftingOperation.java:25]: Also why is this defines as a class member???
This value is not dependent on the state of instances from this class...

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/operations/CraftingOperation.java [hunk 9]
```diff
@@ -35,12 +37,21 @@ public CraftingOperation(@Nonnull ItemStack[] ingredients, @Nonnull ItemStack[]
...
     @Override
     public void addProgress(int num) {
         Validate.isTrue(num > 0, "Progress must be positive.");
-        currentTicks += num;
+
+        /**
+         * Normalize the tickrate to the config value custom-ticker-delay.
+         * This makes the machines run nearly the same speed on changing the delay.
+         */
+        int normalizedTickRate = (int) Math.round((tickRate / 10.0d) * num);
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/operations/CraftingOperation.java:52]:
```suggestion
        int normalizedTickRate = (int) Math.round((tickRate / 10.0D) * num);
```

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]
Lots of changes required.
Many of them are related to formatting.
For reference: Double-values are either displayed as `x.x` or with a capital letters.
Lowercase letters should not be used for explicit number typing.

Also there is no reason for you to store the tick rate as a class member as that shouldn't change on a per-instance-level.

**Comment by simon44556** [comment]
Ok I think that now it should be ok.

I went through all the comments and redid the formating and the config access.

**Comment by svr333** [comment]
Tip for future PRs, you can mark any Requested Change as Resolved, that way it's clear what has been addressed and what wasn't

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/NumberUtils.java [hunk 10]
```diff
@@ -171,9 +170,29 @@ private NumberUtils() {}
...
+        if (normalScale < 0) {
+            //Adjust for positive change
+            seconds = (int) (ticksLeft * (normalScale / 2 + 1.0D));
+        } else if (normalScale > 0) {
+            // Adjust for negative change 
+            seconds = (int) (ticksLeft * (normalScale + 1.0D));
+        } else {
+            // No change 
+            seconds = ticksLeft;
+        }
```
Comment (svr333) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/NumberUtils.java:194]: Not a necessary change but we can get rid of the else bracket by defaulting seconds to ticksLeft

```suggestion
        int seconds = ticksLeft;

        if (normalScale < 0) {
            //Adjust for positive change
            seconds = (int) (ticksLeft * (normalScale / 2 + 1.0D));
        } else if (normalScale > 0) {
            // Adjust for negative change 
            seconds = (int) (ticksLeft * (normalScale + 1.0D));
        }
```

**Review by svr333 — COMMENTED** [review]
PR looks good to me, thanks for your contribution

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/NumberUtils.java (hunk 10 above)
Comment (simon44556) [reply src/main/java/io/github/thebusybiscuit/slimefun4/utils/NumberUtils.java]: Good idea yes.

**Review by svr333 — APPROVED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/operations/CraftingOperation.java [hunk 11]
```diff
@@ -40,7 +40,15 @@ public CraftingOperation(@Nonnull ItemStack[] ingredients, @Nonnull ItemStack[]
     @Override
     public void addProgress(int num) {
         Validate.isTrue(num > 0, "Progress must be positive.");
-        currentTicks += num;
+
+        /**
+         * Normalize the tickrate to the config value custom-ticker-delay.
+         * This makes the machines run nearly the same speed on changing the delay.
+         */
+        int tickRate = Slimefun.getCfg().getInt("URID.custom-ticker-delay");
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/operations/CraftingOperation.java:48]:
```suggestion
        int tickRate = Slimefun.getTickerTask().getTickRate();
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/NumberUtils.java [hunk 12]
```diff
@@ -171,9 +170,26 @@ private NumberUtils() {}
...
 
-    public static @Nonnull String getTimeLeft(int seconds) {
+    public static @Nonnull String getTimeLeft(int ticksLeft) {
         String timeleft = "";
 
+        /**
+         * This is the amount calculated in addProgress(). 
+         * This will be the scale used for timer.
+         */
+        double timeScale = Slimefun.getCfg().getInt("URID.custom-ticker-delay") / 10.0D;
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/NumberUtils.java:180]:
```suggestion
        double timeScale = Slimefun.getTickerTask().getTickRate() / 10.0D;
```

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]
As I mentioned in my last review: Config readouts can be expensive operations!

It is much safer and more efficient to just use the provided getter for reading the cached value instead.

**Review by J3fftw1 — APPROVED** [review]
Will make needed changes so its not a breaking API endpoint.

**Pull request merged**